### PR TITLE
curiosity

### PR DESCRIPTION
**Pull request by zhchaoo**
experiment curiosity, use icm to add intrinsic rewards.
Deepak Pathak, Pulkit Agrawal, Alexei A. Efros and Trevor Darrell. Curiosity-driven Exploration by Self-supervised Prediction. In ICML 2017.

### CODE REVIEW

**Comment by MrSyee** [comment]
Hi, @zhchaoo 
It is great job! I think it'll be better if you make a few change.
1. The graphs of losses for ICM in result part is needed for showing difference between naive DQN and curiosity.
2. In render part, the result gif video is needed, not an error message.

@Curt-Park What do you think? Curiosity is not rainbow series, so it is different from the concept of this tutorial. But it is also one of algorithms using DQN.

**Comment by Curt-Park** [comment]
@MrSyee I think this can be added as an extra topic.
@zhchaoo Thanks for your contribution. Please check @MrSyee 's comment.

**Comment by Curt-Park** [comment]
Colab URL:
https://colab.research.google.com/github/zhchaoo/rainbow-is-all-you-need/blob/feature/curiosity/11.curiosity.ipynb

스크린샷 2021-07-19 오전 12 39 04

File: 11.curiosity.ipynb [hunk 1]
```diff
@@ -0,0 +1,723 @@
+{
```
Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <p>This seems like the most important part in this tutorial.</p><p>Currently, it doesn't look having self-contained information so that people understand it enough by this.</p><p>How about adding more information like mathematical formulation? or the difference from the naive DQN?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #38.            # Forward prediction: predict next state feature, given current state feature and action (one-hot)
</pre><p> </p><p>Suggestion (Google Docstyle):</p><p> </p><p> </p><pre class="ql-syntax" spellcheck="false">"""Forward prediction.

predict next state feature, given current state feature and action (one-hot).
"""
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #39.            pred_s_next = F.relu(self.pred_module1( torch.cat([feature_x, a_vec.float()], dim = -1).detach()))
</pre><p>suggestion (Black Style):</p><p> </p><pre class="ql-syntax" spellcheck="false">pred_s_next = F.relu(
    self.pred_module1(torch.cat([feature_x, a_vec.float()], dim =-1).detach())
)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #59.            self.use_extrinsic = use_extrinsic
</pre><p>Need to add description in the docstring</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #60.            self.intrinsic_scale = intrinsic_scale
</pre><p>Need to add description in the docstring</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #73.            self.icm = ICM(obs_dim, action_dim).to(self.device)
</pre><p>Need to add description in the docstring</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #201.            a_vec = F.one_hot(action, num_classes = self.env.action_space.n).reshape(-1,self.env.action_space.n) # convert action from int to one-hot format
</pre><p>Line is too long. Suggestion (Black style):</p><pre class="ql-syntax" spellcheck="false"># convert action from int to one-hot format
a_vec = F.one_hot(
    action, num_classes=self.env.action_space.n
).reshape(-1, self.env.action_space.n) 
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <p>need to show the result</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (Curt-Park) [inline 11.curiosity.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #37.        def pred(self, feature_x, a_vec):
</pre><p>Need to add type annotation:</p><pre class="ql-syntax" spellcheck="false">feature_x: torch.Tensor, a_vec: torch.Tensor
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Curt-Park/rainbow-is-all-you-need/pull/53/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

**Comment by Curt-Park** [comment]
staled
